### PR TITLE
Add extra test coverage to PKI 

### PR DESCRIPTION
**Pull request by stevendpclark**
* Add a test that validates the behavior of the delete role api
 * Add a test for the crl/rotate api 
 * Rework existing test for fetch raw certs to no longer write directly to storage, instead use the proper apis 
 * Add test for ca_chain fetch api (returning nothing on root)

### CODE REVIEW

File: builtin/logical/pki/backend_test.go [hunk 1]
```diff
@@ -1714,19 +1714,92 @@ func generateRoleSteps(t *testing.T, useCSRs bool) []logicaltest.TestStep {
...
+	})
+	require.NoError(t, err)
+	if resp != nil && resp.IsError() {
+		t.Fatalf("failed read ca/pem, %#v", resp)
+	}
+	// check the raw cert matches the response body
+	if bytes.Compare(resp.Data[logical.HTTPRawBody].([]byte), []byte(rootCaAsPem)) != 0 {
+		t.Fatalf("failed to get raw cert")
+	}
+	// Now issue a short-lived certificate from our pki-external.
```
Comment (kitography) [inline builtin/logical/pki/backend_test.go:1771]: Complete and utter nit: This comment should go after the role generation, or would be cool to remove the new line (1784)

File: builtin/logical/pki/backend_test.go [hunk 2]
```diff
@@ -1714,19 +1714,92 @@ func generateRoleSteps(t *testing.T, useCSRs bool) []logicaltest.TestStep {
...
+	require.NoError(t, err, "error issuing certificate: %v", err)
+	require.NotNil(t, resp, "got nil response from issuing request")
+
+	issueCrtAsPem := resp.Data["certificate"].(string)
+	issuedCrt := parseCert(t, issueCrtAsPem)
+	expectedSerial := certutil.GetHexFormatted(issuedCrt.SerialNumber.Bytes(), ":")
+	expectedCert := []byte(issueCrtAsPem)
 
-	expectedSerial := "17:67:16:b0:b9:45:58:c0:3a:29:e3:cb:d6:98:33:7a:a6:3b:66:c1"
-	expectedCert := []byte("test certificate")
```
Comment (kitography) [inline builtin/logical/pki/backend_test.go:1729]: Why does the "put" below remain?  Wouldn't this be done by issuing the cert?
Also what's the difference between require.NoError and t.Fatal(err)?  They look the same to me.

File: builtin/logical/pki/backend_test.go [hunk 3]
```diff
@@ -3379,6 +3452,126 @@ func TestBackend_AllowedDomainsTemplate(t *testing.T) {
...
+	}
+
+	// Read the role.
+	resp, err = client.Logical().ReadWithContext(ctx, "pki/roles/test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if resp.Data == nil {
+		t.Fatal("response data was nil when it should have contained data")
```
Comment (kitography) [inline builtin/logical/pki/backend_test.go:3506]: nit: default data (makes this more clear without skipping ahead)

File: builtin/logical/pki/backend_test.go [hunk 4]
```diff
@@ -3379,6 +3452,126 @@ func TestBackend_AllowedDomainsTemplate(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	resp, err = client.Logical().ReadWithContext(ctx, "pki/roles/test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if resp != nil {
+		t.Fatalf("response should have been emtpy but was:\n%#v", resp)
```
Comment (kitography) [inline builtin/logical/pki/backend_test.go:3571]: nit: empty

File: builtin/logical/pki/crl_test.go [hunk 5]
```diff
@@ -64,16 +67,8 @@ func TestBackend_CRL_EnableDisable(t *testing.T) {
...
-		crlPem := resp.Data["certificate"].(string)
-		certList, err := x509.ParseCRL([]byte(crlPem))
-		if err != nil {
-			t.Fatal(err)
-		}
-		lenList := len(certList.TBSCertList.RevokedCertificates)
+		certList := getCrlCertificateList(t, client)
+		lenList := len(certList.RevokedCertificates)
 		if lenList != num {
 			t.Fatalf("expected %d, found %d", num, lenList)
```
Comment (kitography) [inline builtin/logical/pki/crl_test.go:73]: nit found %d what? (I presume revoked ceritificates)

**Review by kitography — APPROVED** [review]
Great, beautiful tests fix some major holes, some nits

File: builtin/logical/pki/backend_test.go (hunk 1 above)
Comment (stevendpclark) [reply builtin/logical/pki/backend_test.go]: Addressed.

File: builtin/logical/pki/backend_test.go (hunk 4 above)
Comment (stevendpclark) [reply builtin/logical/pki/backend_test.go]: oops, good eye, addressed.

File: builtin/logical/pki/crl_test.go (hunk 5 above)
Comment (stevendpclark) [reply builtin/logical/pki/crl_test.go]: addressed

File: builtin/logical/pki/backend_test.go (hunk 2 above)
Comment (stevendpclark) [reply builtin/logical/pki/backend_test.go:1729]: Wow I flubbed this one, thanks for catching it. Removed the put and addressed the fallout.

`require.NoError(t, err)` is technically a shortcut to this entire block of code. It's from the [testify 3rd party library](https://github.com/stretchr/testify).
```
if err != nil {
        t.Fatal(err)
}
```

`t.Fatal(err)` fails the test and prints out whatever parameter/string you give it as a parameter, even if err was nil.

File: builtin/logical/pki/backend_test.go (hunk 3 above)
Comment (stevendpclark) [reply builtin/logical/pki/backend_test.go]: addressed.

**Review by cipherboy — APPROVED** [review]

**Pull request merged**